src/include/sof/drivers/dmic.h

```diff
@@ -315,17 +316,33 @@
 #define dmic_irq(dmic) dmic->plat_data.irq
 #define dmic_irq_name(dmic) dmic->plat_data.irq_name
 
-/* DMIC private data */
-struct dmic_pdata {
-	uint16_t enable[DMIC_HW_CONTROLLERS];
-	uint32_t state;
+struct dmic_private_nonshared {
 	struct task dmicwork;
 	int32_t startcount;
 	int32_t gain;
-	int32_t gain_coef;
+};
+
+struct dmic_private_shared {
+	uint16_t enable[DMIC_HW_CONTROLLERS];	/* set_config() on core 0 */
+	uint32_t state;				/* probe() on core 0 */
+	int32_t gain_coef;			/* set_config() on core 0 */
+	int irq;				/* probe() on core 0 */
+};
+
+struct dmic_global_shared {
+	struct sof_ipc_dai_dmic_params prm[DMIC_HW_FIFOS];
+	uint32_t state;
+	int active_fifos_mask;
 	int irq;
 };
 
+/* DMIC private data */
+struct dmic_pdata {
+	struct dmic_global_shared *global;
+	struct dmic_private_nonshared *nonshared;
+	struct dmic_private_shared *shared;
```

Comment:
Whats the difference between ```shared``` and ```global``` ? It's probably very useful to comment the rules of selection (where to put the members) so that new code can be added correctly.

Reply:
I learned that DAI probe() and set_config() always happen in core 0. If the rest of driver runs on another core where the pipeline is then those DAI instance internals need to be shared. The non-shared is stuff that does not need to carry from probe or set config to other DAI functions. Se we need two categories of shared data.

Reply:
@lgirdwood This patch need to be rebased after DSP panic fix #4313 merge. It shares the same pause/stop problem as git main currently.

Reply:
> I learned that DAI probe() and set_config() always happen in core 0. If the rest of driver runs on another core where the pipeline is then those DAI instance internals need to be shared. The non-shared is stuff that does not need to carry from probe or set config to other DAI functions. Se we need two categories of shared data.

ok, but we cant always assume probe() and config() will al;ways be core 0 - simpler to just have shared(global) and nonshared(instance). 3 categories makes it more confusing.

Reply:
@lgirdwood It's not possible to merge like you propose. The shared data that is not global is different for DMIC DAI instances. The only way to merge is to have shared global (common data for DMIC48k and DMIC16k) and shared within per DAI instance. Which way do you prefer?

Reply:
what about 
```c
struct dmic_shared {
    struct global glb;
    struct instance dmic[num_dmics];
};
```

Reply:
@lgirdwood @singalsu IIUC "global" means "common for all DMICs," whereas "shared" means "potentially accessed by multiple cores." In this case also "global" is static data, whereas "shared" is dynamically allocated

Reply:
@lgirdwood I'm missing maybe something but why expose something to other DMIC instance that is not needed?

Reply:
@lgirdwood @lyakh I didn't change this yet in this update. Need to confirm me about benefit of doing other way! The "nonshared" could be eliminated from this PR if preferred.

Reply:
ok, latest code looks better.

Reply:
@lgirdwood It's very problematic to make the driver code clean if there are no guarantee whether the functions are called from same core or different. @lyakh had a proposal that we should do a cache invalidate after probe and set_config and then driver would be cleaner. Ideally this invalidate would be done at a higher lever where the decision is made to call the code from different cores. In practise, this is probably hard to do without invalidating the whole cache. So maybe pragmatic would be to add the cache invalidte to driver probe() and set_config(). Then Seppo could have just two sections (global = shared by separate instances of this driver, private = only used by single driver instance, with assumption all accesses are done by the same core). Makes sense?

Reply:
I really fear when we make "who is calling us where where" based assumptions we will end up repeating this debug. TBH, it's not going to be clean - the cleanest is to allocate all dmic driver data using the uncache alias. It's not used at copy() anyway.

Reply:
It's fragile, true. In this case it would seem none of this data is accessed on the hot-path, so maybe for this case allocating all with uncached alias is the best. We don't lose anything, and it's robust. For other drivers, this couldb be more complicated if they cannot assume anything about the core, but need to access instance data from a hot path.